File: sdk/elixir/lib/bunqueue/flow_producer.ex
defmodule Bunqueue.FlowProducer do
  @moduledoc """
  Plans flows locally and commits each tree or chain with one atomic `PUSHF`.

  Job identifiers and reciprocal topology links are fully resolved before the
  broker sees any job. Returned nodes are built from authoritative broker
  snapshots, not reconstructed request data.
  """

  alias Bunqueue.{Connection, FlowPlanner, FlowSnapshots, Job}

  @enforce_keys [:connection]
  defstruct [:connection, owns_connection: true]

  def new(options \\ []) do
    {:ok, connection} = Connection.start_link(options)
    %__MODULE__{connection: connection}
  end

  def with_connection(connection),
    do: %__MODULE__{connection: connection, owns_connection: false}

  @spec add(t(), map()) :: {:ok, map()} | {:error, Exception.t()}
  def add(producer, flow) do
    protect(fn ->
      plan = FlowPlanner.plan_tree(flow)

      with {:ok, snapshots} <- commit(producer.connection, plan.jobs) do
        {:ok, build_node(plan.root, snapshots, producer.connection)}
      end
    end)
  end

  @spec add_chain(t(), [map()]) :: {:ok, [String.t()]} | {:error, Exception.t()}
  def add_chain(producer, steps) do
    protect(fn ->
      plan = FlowPlanner.plan_chain(steps)

      with {:ok, _snapshots} <- commit(producer.connection, plan.jobs) do
        {:ok, plan.ids}
      end
    end)
  end

  def close(%{owns_connection: true, connection: connection}), do: Connection.close(connection)
  def close(_producer), do: :ok

  defp commit(_connection, []), do: {:ok, %{}}

  defp commit(connection, jobs) do
    case Connection.call(connection, %{"cmd" => "PUSHF", "jobs" => jobs}) do
      {:ok, response} ->
        snapshots = get_in(response, ["data", "jobs"])
        {:ok, FlowSnapshots.index!(jobs, snapshots)}

      {:error, error} ->
        {:error, error}
    end
  end

  defp build_node(node, snapshots, connection) do
    %{
      job: snapshots |> Map.fetch!(node.id) |> Job.from_wire(connection),
      children: Enum.map(node.children, &build_node(&1, snapshots, connection))
    }
  end

  defp protect(fun) do
    fun.()
  rescue
    error -> {:error, error}
  end

  @type t :: %__MODULE__{
          connection: GenServer.server(),
          owns_connection: boolean()
        }
end

File: sdk/php/src/FlowProducer.php
<?php

declare(strict_types=1);

namespace Bunqueue;

use Bunqueue\Exception\CommandException;
use Bunqueue\Flow\Planner;
use Bunqueue\Flow\SnapshotValidator;

/**
 * FlowProducer: parent/child job trees and sequential chains.
 *
 * Every ID and edge is resolved locally, then the complete graph is committed
 * by one broker-side PUSHF operation. No partially linked flow is observable.
 */
final class FlowProducer
{
    public readonly Connection $connection;
    private readonly bool $ownsConnection;
    /** @var \Closure(array<string, mixed>): array<string, mixed> */
    private \Closure $call;

    /** @param array{host?: string, port?: int, token?: string, tls?: bool|array} $options */
    public function __construct(array $options = [], ?Connection $connection = null)
    {
        $this->connection = $connection ?? new Connection($options);
        $this->ownsConnection = $connection === null;
        $this->call = fn (array $command): array => $this->connection->call($command);
    }

    /**
     * Add a flow tree: `['name', 'queueName', 'data', 'opts', 'children' => [...]]`.
     */
    public function add(array $flow): FlowNode
    {
        $plan = (new Planner())->planTree($flow);
        $snapshots = $this->commit($plan['jobs']);
        return $this->buildNode($plan['root'], $snapshots);
    }

    /**
     * Sequential chain: step[0] -> step[1] -> ... (each depends on the previous).
     *
     * @param list<array<string, mixed>> $steps
     * @return list<string> created job ids in chain order
     */
    public function addChain(array $steps): array
    {
        if ($steps === []) {
            return [];
        }
        $plan = (new Planner())->planChain($steps);
        $this->commit($plan['jobs']);
        return $plan['ids'];
    }

    /**
     * Reconstruct a flow tree from a root job id. A removed child (or root)
     * yields null / a partial tree instead of throwing; a visited set guards
     * against cycles in childrenIds.
     */
    public function getFlow(string $jobId, ?int $depth = null): ?FlowNode
    {
        $visited = [];
        return $this->fetchNode($jobId, $depth, $visited);
    }

    public function close(): void
    {
        if ($this->ownsConnection) {
            $this->connection->close();
        }
    }

    // ------------------------------------------------------------ internals

    /**
     * @param list<array{id: string, queue: string, input: array<string, mixed>}> $jobs
     * @return array<string, array<string, mixed>>
     */
    private function commit(array $jobs): array
    {
        $response = ($this->call)(['cmd' => 'PUSHF', 'jobs' => $jobs]);
        $data = $response['data'] ?? null;
        $rawJobs = \is_array($data) ? ($data['jobs'] ?? null) : null;
        return SnapshotValidator::validate($jobs, $rawJobs);
    }

    /**
     * @param array{id: string, children: list<array>} $planned
     * @param array<string, array<string, mixed>> $snapshots
     */
    private function buildNode(array $planned, array $snapshots): FlowNode
    {
        $children = array_map(
            fn (array $child): FlowNode => $this->buildNode($child, $snapshots),
            $planned['children'],
        );
        return new FlowNode(
            new Job($snapshots[$planned['id']], $this->connection),
            $children,
        );
    }

    /** @param array<string, true> $visited */
    private function fetchNode(string $jobId, ?int $depth, array &$visited): ?FlowNode
    {
        if (isset($visited[$jobId])) {
            return null; // cycle guard
        }
        $visited[$jobId] = true;
        try {
            $response = $this->connection->call(['cmd' => 'GetJob', 'id' => $jobId]);
        } catch (CommandException $e) {
            if (str_contains(strtolower($e->getMessage()), 'not found')) {
                return null; // root or a since-removed child -> skip
            }
            throw $e;
        }
        $raw = $response['job'] ?? null;
        if (!\is_array($raw)) {
            return null;
        }
        $job = new Job($raw, $this->connection);
        $children = [];
        if ($depth === null || $depth > 0) {
            $nextDepth = $depth === null ? null : $depth - 1;
            foreach ($job->childrenIds() as $childId) {
                $child = $this->fetchNode($childId, $nextDepth, $visited);
                if ($child !== null) {
                    $children[] = $child;
                }
            }
        }
        return new FlowNode($job, $children);
    }

}
